File: go/internal/piicore/pii.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

// Package piicore contains the stateless PII sanitization engine shared by
// the top-level telemetry package and the logger sub-package.  All mutable
// state (rule lists, hooks, custom patterns) lives in the callers; this
// package only provides pure functions. The one exception is the hash
// canonicalizer hook (SetHashCanonicalizer), which exists because the RFC
// 8785 serializer lives in the root package and the root package imports
// this one.
package piicore

import (
	"crypto/sha256"
	"fmt"
	"regexp"
	"sort"
	"strings"
	"sync/atomic"
)

// PIIRule defines a rule for sanitizing a specific field path.
//
// TruncateTo only matters in truncate mode. Go's zero value means "unset":
// the root package normalises 0 to DefaultTruncateTo when the rule is
// registered, so a rule built without a limit truncates to 8 code points like
// every other SDK. A negative limit is clamped to 0 at apply time, which
// keeps nothing but the suffix — never an error, never the whole value.
type PIIRule struct {
	Path       []string
	Mode       string
	TruncateTo int
}

// DefaultTruncateTo is the truncate-mode limit used when a rule carries none.
const DefaultTruncateTo = 8

// PII mode constants.
const (
	PIIModeRedact   = "redact"
	PIIModeDrop     = "drop"
	PIIModeHash     = "hash"
	PIIModeTruncate = "truncate"
	PIIModePass     = "pass"
)

// Exported sentinel values used by both callers.
const (
	Redacted         = "***"
	TruncationSuffix = "..."
	DefaultMaxDepth  = 8
)

// DefaultSensitiveKeys lists case-insensitive exact-match key names that are
// redacted automatically even when no custom rule matches.
var DefaultSensitiveKeys = map[string]struct{}{
	"password":       {},
	"passwd":         {},
	"secret":         {},
	"token":          {},
	"api_key":        {},
	"apikey":         {},
	"auth":           {},
	"authorization":  {},
	"credential":     {},
	"private_key":    {},
	"ssn":            {},
	"credit_card":    {},
	"creditcard":     {},
	"cvv":            {},
	"pin":            {},
	"account_number": {},
	"cookie":         {},
}

// BuiltinSecretPatterns are the compiled regexps checked against every string // pragma: allowlist secret
// value when no custom rule matches. Patterns are sourced from the generated file.
var BuiltinSecretPatterns = generatedSecretPatterns // pragma: allowlist secret

// ApplyRule returns true if rule.Path matches path (supports '*' wildcards).
func ApplyRule(rule PIIRule, path []string) bool {
	if len(rule.Path) != len(path) {
		return false
	}
	for i, seg := range rule.Path {
		if seg != "*" && seg != path[i] {
			return false
		}
	}
	return true
}

// hashCanonicalizer renders a non-string value to the text hash mode digests.
//
// The contract is the RFC 8785 canonical JSON of the value — the same
// serializer the redaction receipts use — so that every SDK hashes true as
// "true", nil as "null" and an object key-sorted. That serializer lives in the
// root package, which imports this one, so it is installed here at init
// through SetHashCanonicalizer rather than imported. Until one is installed
// (this package's own tests, or a binary that links only a sub-package) the
// fallback is fmt's %v rendering.
var hashCanonicalizer atomic.Pointer[func(any) string]

// SetHashCanonicalizer installs the renderer hash mode uses for non-string
// values. Passing nil restores the %v fallback.
func SetHashCanonicalizer(fn func(any) string) {
	if fn == nil {
		hashCanonicalizer.Store(nil)
		return
	}
	hashCanonicalizer.Store(&fn)
}

// hashInput returns the bytes-as-text that hash mode digests for value: the
// string itself, or the canonical rendering of anything else.
func hashInput(value any) string {
	if s, ok := value.(string); ok {
		return s
	}
	if fn := hashCanonicalizer.Load(); fn != nil {
		return (*fn)(value)
	}
	return fmt.Sprintf("%v", value)
}

// ApplyMode applies the given mode to value and returns (result, should_drop).
//
// Truncate counts Unicode code points, never bytes or UTF-16 units, so an
// astral character is kept or dropped whole. A limit of 0 yields exactly the
// suffix; a negative limit is clamped to 0 rather than panicking on the slice.
func ApplyMode(value any, mode string, truncateTo int) (any, bool) {
	switch mode {
	case PIIModeDrop:
		return nil, true
	case PIIModeHash:
		sum := sha256.Sum256([]byte(hashInput(value)))
		return fmt.Sprintf("%x", sum)[:12], false
	case PIIModeTruncate:
		s := fmt.Sprintf("%v", value)
		limit := max(truncateTo, 0)
		runes := []rune(s)
		if len(runes) > limit {
			return string(runes[:limit]) + TruncationSuffix, false
		}
		return s, false
	default:
		return Redacted, false
	}
}

// IsDefaultSensitiveKey returns true if key exactly matches a default
// sensitive key name, case-insensitively.
func IsDefaultSensitiveKey(key string) bool {
	_, ok := DefaultSensitiveKeys[strings.ToLower(key)]
	return ok
}

// DetectSecretInValue returns true if s matches any built-in secret pattern // pragma: allowlist secret
// or any of the caller-supplied custom patterns.
// customPatterns may be nil.
func DetectSecretInValue(s string, customPatterns map[string]*regexp.Regexp) bool { // pragma: allowlist secret
	return len(secretSpans(s, customPatterns)) > 0
}

// pathMinSegments is how many slash-separated parts a span needs before its
// shape is considered path-like.
const pathMinSegments = 3

// looksLikePath reports whether a matched span is a filesystem path rather
// than a secret.
//
// The long_base64 pattern is [A-Za-z0-9+/]{40,} and "/" belongs to the base64
// alphabet, so any deep path of unpunctuated segments matched it:
// /home/deploy/apps/production/current/lib/service is 48 characters of pure
// base64 alphabet holding no secret. Narrowing the charset is not the fix —
// dropping "/" costs 44% of detections on 32-byte secrets, since a 44-char
// base64 string containing one slash cannot be told from a path by charset.
//
// Shape separates them: a path carries several short all-lowercase words
// (usr, local, lib), which random base64 effectively never produces — a
// 20-character all-lowercase run has probability (26/64)^20, about 1e-8.
func looksLikePath(span string) bool {
	segments := make([]string, 0, 8)
	for _, seg := range strings.Split(span, "/") {
		if seg != "" {
			segments = append(segments, seg)
		}
	}
	if len(segments) < pathMinSegments {
		return false
	}
	wordy := 0
	for _, seg := range segments {
		if isLowerAlpha(seg) {
			wordy++
		}
	}
	return wordy*2 >= len(segments)
}

func isLowerAlpha(s string) bool {
	for _, r := range s {
		if r < 'a' || r > 'z' {
			return false
		}
	}
	return s != ""
}

// secretSpans returns every secret-looking span in s, each widened to its
// whitespace-delimited token, sorted and coalesced.
//
// Every pattern is scanned across the WHOLE value, not stopped at its first
// match, and every pattern is tried even after one has hit. Skipping either
// leaks:
//
//   - Stopping a pattern at its first match let a path shadow a real secret.
//     long_base64 matches a path first; suppressing that match as path-shaped
//     moved the scan to the next pattern, and long_base64 is the last one, so
//     the credential behind the path was never looked for at all.
//   - Stopping at the first pattern to hit left a field's second and third
//     secrets in the log, which whole-value blanking used to cover for free.
//
// Collecting every span also makes the result independent of map iteration
// order, which Go randomises — with one span the chosen secret varied between
// runs and diverged from the other four runtimes.
//
// FindStringIndex runs first as a fast path: a clean value, which is nearly
// every log field, pays one scan per pattern and allocates nothing, because
// FindAllStringIndex is only reached once a pattern is known to match.
func secretSpans(s string, customPatterns map[string]*regexp.Regexp) [][2]int {
	if len(s) < MinSecretLength { // pragma: allowlist secret
		return nil
	}
	var spans [][2]int
	collect := func(re *regexp.Regexp) {
		if re.FindStringIndex(s) == nil {
			return
		}
		for _, loc := range re.FindAllStringIndex(s, -1) {
			// A registered pattern that can match the empty string carries no
			// secret; widening a zero-length match to its token would redact a
			// word for nothing.
			if loc[0] == loc[1] {
				continue
			}
			if !looksLikePath(s[loc[0]:loc[1]]) {
				spans = append(spans, expandToToken(s, loc[0], loc[1]))
			}
		}
	}
	for _, re := range BuiltinSecretPatterns {
		collect(re)
	}
	for _, re := range customPatterns {
		collect(re)
	}
	return mergeSpans(spans)
}

// expandToToken widens a match to its whitespace-delimited token.
//
// Redacting the literal match alone can leave part of a credential behind:
// the jwt pattern matches header.payload, and a JWT has THREE dot-separated
// parts, so the signature would survive. Whitespace is the boundary a secret
// cannot cross without ceasing to be one token.
func expandToToken(s string, start, end int) [2]int {
	for start > 0 && !isSpaceByte(s[start-1]) {
		start--
	}
	for end < len(s) && !isSpaceByte(s[end]) {
		end++
	}
	return [2]int{start, end}
}

// mergeSpans sorts spans and coalesces overlaps so each region is replaced
// once. Two patterns can match the same credential — long_base64 and jwt both
// hit a JWT — and after widening they overlap exactly, which would otherwise
// emit "******".
func mergeSpans(spans [][2]int) [][2]int {
	if len(spans) < 2 {
		return spans
	}
	sort.Slice(spans, func(i, j int) bool { return spans[i][0] < spans[j][0] })
	merged := spans[:1]
	for _, span := range spans[1:] {
		last := &merged[len(merged)-1]
		if span[0] <= last[1] {
			if span[1] > last[1] {
				last[1] = span[1]
			}
			continue
		}
		merged = append(merged, span)
	}
	return merged
}

// RedactSecretSpans replaces every secret-looking token of s, leaving the rest
// of the string readable.
//
// Every span goes, not just the first: whole-value blanking removed a field's
// second and third credentials for free, and scoping redaction to a token
// silently dropped that guarantee.
func RedactSecretSpans(s string, customPatterns map[string]*regexp.Regexp) string {
	spans := secretSpans(s, customPatterns)
	if len(spans) == 0 {
		return s
	}
	return replaceSpans(s, spans)
}

// replaceSpans swaps each span for the redaction sentinel.
func replaceSpans(s string, spans [][2]int) string {
	var out strings.Builder
	prev := 0
	for _, span := range spans {
		out.WriteString(s[prev:span[0]])
		out.WriteString(Redacted)
		prev = span[1]
	}
	out.WriteString(s[prev:])
	return out.String()
}

func isSpaceByte(b byte) bool {
	return b == ' ' || b == '\t' || b == '\n' || b == '\r' || b == '\v' || b == '\f'
}

// FireReceiptHook calls hook if non-nil.
func FireReceiptHook(hook func(string, string, any), fieldPath, action string, original any) {
	if hook != nil {
		hook(fieldPath, action, original)
	}
}

// ShallowCopy returns a shallow copy of m.
func ShallowCopy(m map[string]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

// SanitizeMap copies the map, recursively sanitizing values.
// receiptHook may be nil; customPatterns may be nil.
func SanitizeMap(
	m map[string]any,
	path []string,
	rules []PIIRule,
	depth int,
	receiptHook func(string, string, any),
	customPatterns map[string]*regexp.Regexp,
) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		childPath := append(path, k) //nolint:gocritic
		sanitized, drop := SanitizeValue(k, v, childPath, rules, depth, receiptHook, customPatterns)
		if !drop {
			out[k] = sanitized
		}
	}
	return out
}

// SanitizeSlice copies the slice, recursively sanitizing each element.
func SanitizeSlice(
	s []any,
	path []string,
	rules []PIIRule,
	depth int,
	receiptHook func(string, string, any),
	customPatterns map[string]*regexp.Regexp,
) []any {
	out := make([]any, 0, len(s))
	for _, item := range s {
		if inner, ok := item.(map[string]any); ok {
			out = append(out, SanitizeMap(inner, path, rules, depth, receiptHook, customPatterns))
		} else {
			sanitized, drop := SanitizeValue("", item, path, rules, depth, receiptHook, customPatterns)
			if !drop {
				out = append(out, sanitized)
			}
		}
	}
	return out
}

// SanitizeValue applies custom rules, then default key detection, to a single value.
// Returns (sanitized value, should_drop).
func SanitizeValue(
	key string,
	value any,
	path []string,
	rules []PIIRule,
	depth int,
	receiptHook func(string, string, any),
	customPatterns map[string]*regexp.Regexp,
) (any, bool) {
	// Apply custom rules first.
	for _, rule := range rules {
		if ApplyRule(rule, path) {
			FireReceiptHook(receiptHook, strings.Join(path, "."), rule.Mode, value)
			return ApplyMode(value, rule.Mode, rule.TruncateTo)
		}
	}

	// Apply default sensitive key detection.
	if IsDefaultSensitiveKey(key) {
		FireReceiptHook(receiptHook, key, PIIModeRedact, value)
		return Redacted, false
	}

	// Scan string values for known secret patterns. One scan, not two:
	// detecting and then redacting ran the whole pattern sweep twice for every
	// value carrying a credential.
	if str, ok := value.(string); ok {
		if spans := secretSpans(str, customPatterns); len(spans) > 0 {
			FireReceiptHook(receiptHook, key, PIIModeRedact, value)
			// Span-scoped: only the credential tokens go, the message stays.
			return replaceSpans(str, spans), false
		}
	}

	// Recurse into nested structures if depth allows.
	if depth <= 1 {
		return value, false
	}
	switch typed := value.(type) {
	case map[string]any:
		return SanitizeMap(typed, path, rules, depth-1, receiptHook, customPatterns), false
	case []any:
		return SanitizeSlice(typed, path, rules, depth-1, receiptHook, customPatterns), false
	}
	return value, false
}

File: src/provide/telemetry/pii.py
# SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

"""PII policy engine with nested traversal support."""

from __future__ import annotations

__all__ = [
    "MaskMode",
    "PIIRule",
    "get_pii_rules",
    "get_secret_patterns",
    "register_pii_rule",
    "register_secret_pattern",
    "replace_pii_rules",
    "sanitize_payload",
]

import hashlib
import re as _re
import threading
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, Literal, cast

from provide.telemetry._secret_patterns_generated import MIN_SECRET_LENGTH as _MIN_SECRET_LENGTH
from provide.telemetry._secret_patterns_generated import PATTERNS as _RAW_SECRET_PATTERNS

MaskMode = Literal["drop", "redact", "hash", "truncate"]


@dataclass(frozen=True, slots=True)
class PIIRule:
    path: tuple[str, ...]
    mode: MaskMode = "redact"
    truncate_to: int = 8


_SECRET_PATTERNS: tuple[tuple[str, _re.Pattern[str]], ...] = tuple(
    (name, _re.compile(pattern)) for name, pattern in _RAW_SECRET_PATTERNS
)

_custom_secret_patterns: list[tuple[str, _re.Pattern[str]]] = []

# ReDoS safety cap: values longer than this are never scanned for secret
# patterns.  API-shaped secrets are short (tens to low hundreds of chars); a
# >8 KiB string almost certainly isn't a key, and scanning it exposes the
# regex engine to pathological catastrophic-backtracking inputs.
_MAX_SECRET_SCAN_LENGTH = 8192


def register_secret_pattern(name: str, pattern: _re.Pattern[str]) -> None:
    """Register a custom secret detection pattern.

    If *name* already exists, the previous pattern is replaced (deduplication).
    The *name* is for diagnostics only and is not used during matching.
    """
    with _lock:
        for idx, (existing_name, _pat) in enumerate(_custom_secret_patterns):
            if existing_name == name:
                _custom_secret_patterns[idx] = (name, pattern)
                return
        _custom_secret_patterns.append((name, pattern))


def get_secret_patterns() -> tuple[tuple[str, _re.Pattern[str]], ...]:
    """Return all secret patterns (built-in and custom)."""
    with _lock:
        return _SECRET_PATTERNS + tuple(_custom_secret_patterns)


# A slash-separated segment count and word shape that says "filesystem path".
# The long_base64 pattern is [A-Za-z0-9+/]{40,}, and "/" is in the base64
# alphabet, so any deep path of unpunctuated segments matches it:
# /home/deploy/apps/production/current/lib/service is 48 characters of pure
# base64 alphabet and contains no secret at all. Narrowing the charset is not
# the fix -- measured, dropping "/" costs 44% of detections on 32-byte secrets,
# the most common size there is, because a 44-char base64 string containing one
# slash is indistinguishable from a path BY CHARSET.
#
# Shape distinguishes them. A path has several short all-lowercase words (usr,
# local, lib); random base64 essentially never does -- a 20-character
# all-lowercase run has probability (26/64)^20, about 1e-8. Measured over 20k
# random secrets per size: detection stays 100% and 0.00-0.06% are suppressed.
_PATH_MIN_SEGMENTS = 3

# Token boundaries for _expand_to_token. `\S` is the exact complement of
# str.isspace() across the whole BMP, so switching from a hand-rolled
# character walk to these changed no behaviour.
_RUN_BEFORE_END = _re.compile(r"\S*\Z")
_RUN_FROM_POS = _re.compile(r"\S*")


def _looks_like_path(span: str) -> bool:
    """Return True if *span* has the shape of a filesystem path, not a secret."""
    segments = [segment for segment in span.split("/") if segment]
    if len(segments) < _PATH_MIN_SEGMENTS:
        return False
    wordy = sum(1 for segment in segments if segment.isalpha() and segment.islower())
    return wordy * 2 >= len(segments)


def _secret_spans(value: str) -> list[tuple[int, int]]:
    """Return every secret-looking span in *value*, merged and in order.

    Every pattern is scanned across the WHOLE value, not stopped at its first
    match, and every pattern is tried even after one has hit. Both halves
    matter, and skipping either leaks:

    * Stopping a pattern at its first match let a path shadow a real secret.
      long_base64 matches a path first; suppressing that match as path-shaped
      moved the scan to the next pattern, and long_base64 is the last one, so
      the credential behind the path was never looked for at all.
    * Stopping at the first *pattern* to hit left a field's second and third
      secrets in the log, which whole-value blanking used to cover for free.

    The fast path is search(): a clean value -- nearly every log field -- pays
    one search() per pattern and allocates nothing, because finditer() is only
    entered once a pattern is known to match. Measured, that keeps the clean
    case at +0.3-10% rather than the +92% an unconditional finditer() costs.
    """
    if len(value) < _MIN_SECRET_LENGTH:
        return []
    if len(value) > _MAX_SECRET_SCAN_LENGTH:
        return []
    with _lock:
        custom = list(_custom_secret_patterns)
    # Two loops rather than one over (*builtin, *custom): a clean value is
    # nearly every log field, and building that tuple per call cost more than
    # the duplicated loop body. The span list is likewise only allocated once
    # something has actually matched.
    spans: list[tuple[int, int]] | None = None
    for patterns in (_SECRET_PATTERNS, custom):
        for _name, pattern in patterns:
            # search() is only a presence check -- it is what keeps a clean
            # value cheap. Once a pattern is known to match, finditer() walks
            # the whole value from the start so no later match is missed.
            if pattern.search(value) is None:
                continue
            if spans is None:
                spans = []
            for match in pattern.finditer(value):
                matched = match.group(0)
                # A registered pattern that can match the empty string carries
                # no secret; widening a zero-length match to its token would
                # redact a word for nothing.
                if not matched:
                    continue
                if not _looks_like_path(matched):
                    spans.append(_expand_to_token(value, *match.span()))
    if spans is None:
        return []
    return _merge_spans(spans)


def _merge_spans(spans: list[tuple[int, int]]) -> list[tuple[int, int]]:
    """Sort and coalesce overlapping spans so each region is replaced once.

    Two patterns can match the same credential -- long_base64 and jwt both hit
    a JWT -- and after widening to the token they overlap exactly. Replacing
    each separately would emit "******".
    """
    merged: list[tuple[int, int]] = []
    for start, end in sorted(spans):
        # `<` and `<=` are indistinguishable here, so the `<` mutant is
        # equivalent: every span has been widened to a whitespace-delimited
        # token, so one span's start is at least one whitespace character past
        # the previous span's end. start == previous end cannot occur.
        if merged and start <= merged[-1][1]:  # pragma: no mutate — see above
            previous_start, previous_end = merged[-1]
            merged[-1] = (previous_start, max(previous_end, end))
        else:
            merged.append((start, end))
    return merged


def _expand_to_token(value: str, start: int, end: int) -> tuple[int, int]:
    """Widen a match to its whitespace-delimited token.

    Redacting only the literal match can leave part of a secret behind: the
    jwt pattern matches header.payload, and a JWT has THREE dot-separated
    parts, so the signature would survive untouched. Whitespace is the
    boundary a secret cannot cross without ceasing to be one token, so
    widening to it means a partial match still removes the whole credential
    while the words around it stay readable.

    The boundaries are found by anchored match rather than by walking an
    index. Index arithmetic here was the wrong tool twice over: a ``while``
    loop is one mutated assignment away from never terminating, and the
    ``range`` form that replaced it carried four separate constants whose
    off-by-one variants mostly cannot be told apart from the outside. The
    regexes say the thing directly -- the run of non-whitespace ending where
    the match begins, and the run starting where it ends.
    """
    # Both patterns can match empty, so neither call can return None -- but
    # nothing in the signature says so, and the repo runs two type checkers
    # that each need telling separately.
    return (
        _RUN_BEFORE_END.search(value, 0, start).start(),  # type: ignore[union-attr]  # ty: ignore[unresolved-attribute]
        _RUN_FROM_POS.match(value, end).end(),  # type: ignore[union-attr]  # ty: ignore[unresolved-attribute]
    )


def redact_secret_spans(value: str) -> str:
    """Replace every secret-looking SPAN of *value*, leaving the rest.

    Blanking the whole field destroys context that was never suspect: a
    remediation string that happened to contain a long path became "***" and
    stopped telling anyone what to run. For a value that IS a secret the span
    covers all of it, so that case is unchanged.

    Every span is replaced, not just the first. Whole-value blanking removed a
    field's second and third credentials for free, and scoping redaction to a
    token silently dropped that guarantee.
    """
    spans = _secret_spans(value)
    if not spans:
        return value
    return _replace_spans(value, spans)


def _replace_spans(value: str, spans: list[tuple[int, int]]) -> str:
    """Swap each span for the redaction sentinel.

    Walks the spans back to front so each replacement leaves the indices of
    the ones still to come untouched. Carrying a cursor forward instead needs
    a zero sentinel whose only wrong value, None, slices identically -- an
    unkillable mutant guarding nothing.
    """
    redacted = value
    for start, end in reversed(spans):
        redacted = redacted[:start] + _REDACTED + redacted[end:]
    return redacted


def _redact_if_secret(value: str) -> str | None:
    """Redacted form of *value*, or None when it holds no secret.

    One scan where the sanitize path used to do two. Asking
    _detect_secret_in_value and then redact_secret_spans ran the whole pattern
    sweep twice for every value carrying a credential, which got measurably
    worse once the sweep started collecting every match instead of stopping at
    the first.
    """
    spans = _secret_spans(value)
    if not spans:
        return None
    return _replace_spans(value, spans)


def _detect_secret_in_value(value: str) -> bool:
    """Return True if value holds a known secret.

    Shares _secret_spans with redaction rather than running its own loop. When
    the two disagreed the value was flagged and then not fully cleaned, which
    is exactly how a secret sitting behind a filesystem path escaped: this
    returned False because the only pattern that could match it had already
    been consumed by the path.

    The length guards and the ReDoS scan cap live in _secret_spans.
    """
    return bool(_secret_spans(value))


_DEFAULT_SENSITIVE_KEYS = {
    "password",
    "passwd",
    "secret",
    "token",
    "api_key",
    "apikey",
    "auth",
    "authorization",
    "credential",
    "private_key",
    "ssn",
    "credit_card",
    "creditcard",
    "cvv",
    "pin",
    "account_number",
    "cookie",
}
_lock = threading.Lock()
_rules: list[PIIRule] = []

# Governance hooks — set by classification.py / receipts.py if present.
# None = feature not loaded (zero overhead).
_classification_hook: Callable[[str, Any], str | None] | None = None
_receipt_hook: Callable[[str, str, Any], None] | None = None
# Set by classification.py when rules are registered; takes label → action string.
_policy_hook: Callable[[str], str] | None = None


def replace_pii_rules(rules: list[PIIRule]) -> None:
    with _lock:
        _rules.clear()
        _rules.extend(rules)


def register_pii_rule(rule: PIIRule) -> None:
    with _lock:
        _rules.append(rule)


def get_pii_rules() -> tuple[PIIRule, ...]:
    with _lock:
        return tuple(_rules)


_REDACTED = "***"
_TRUNCATION_SUFFIX = "..."


def _mask(value: Any, mode: MaskMode, truncate_to: int) -> Any:
    if mode == "drop":
        return None
    if mode == "redact":
        return _REDACTED
    if mode == "hash":
        if isinstance(value, str):
            text = value
        else:
            # Non-strings hash their RFC 8785 canonical JSON so every SDK spells
            # booleans, null, numbers and nested values identically. Imported
            # lazily: receipts imports this module.
            from provide.telemetry.receipts import canonical_json

            text = canonical_json(value)
        # Codec lookup is case-insensitive, so an "UTF-8" mutation is equivalent.
        encoded = text.encode("utf-8")  # pragma: no mutate — codec alias; 'UTF-8' selects the identical codec
        return hashlib.sha256(
            encoded
        ).hexdigest()[
            :12
        ]  # pragma: no mutate — 12-char hash prefix is the PII hash-mode contract; exact value asserted in hash-mode tests
    text = str(value)
    limit = max(0, truncate_to)
    if len(text) <= limit:
        return text
    return text[:limit] + _TRUNCATION_SUFFIX


def _match(path: tuple[str, ...], target: tuple[str, ...]) -> bool:
    if len(path) != len(target):
        return False
    # The length guard above means zip can never see a ragged pair, so every
    # mutation of `strict` here is provably equivalent. Kept as True to document
    # the invariant. The pragma needs a single-line statement to take effect.
    pairs = zip(path, target, strict=True)  # pragma: no mutate — length guard above means zip is never ragged
    return all(part == "*" or part == elem for part, elem in pairs)


def _apply_rule(
    node: Any,
    rule: PIIRule,
    current_path: tuple[str, ...] = (),
    depth: int = 0,  # pragma: no mutate — recursion-depth default; call sites always start at 0
    receipt_hook: Callable[[str, str, Any], None] | None = None,
) -> Any:
    if depth >= 32:  # hard safety limit
        return node
    if isinstance(node, dict):
        output: dict[str, Any] = {}
        for key, value in node.items():
            child_path = (*current_path, key)
            if _match(rule.path, child_path):
                masked = _mask(value, rule.mode, rule.truncate_to)
                if masked is not None:
                    output[key] = masked
                if receipt_hook is not None:
                    receipt_hook(".".join(child_path), rule.mode, value)
            else:
                output[key] = _apply_rule(value, rule, child_path, depth=depth + 1, receipt_hook=receipt_hook)
        return output
    if isinstance(node, list):
        return [
            _apply_rule(item, rule, (*current_path, "*"), depth=depth + 1, receipt_hook=receipt_hook) for item in node
        ]  # pragma: no mutate — list-comp structure; traversal asserted by nested-list PII rule tests
    return node


def _path_has_rule(rule_paths: frozenset[tuple[str, ...]], child_path: tuple[str, ...]) -> bool:
    """Return True if any rule path matches child_path via _match()."""
    return any(_match(rp, child_path) for rp in rule_paths)


def _apply_default_sensitive_key_redaction(
    node: Any,
    original: Any,
    depth: int = 0,  # pragma: no mutate — recursion-depth default; call sites always start at 0
    max_depth: int = 8,  # pragma: no mutate — default max_depth is overridden by live runtime config at every call
    receipt_hook: Callable[[str, str, Any], None] | None = None,
    rule_targeted_paths: frozenset[tuple[str, ...]] | None = None,
    _current_path: tuple[str, ...] = (),
) -> Any:
    if depth >= max_depth:
        return node
    if rule_targeted_paths is None:
        rule_targeted_paths = frozenset()
    if isinstance(node, dict) and isinstance(original, dict):
        output: dict[str, Any] = {}
        for key, value in node.items():
            orig_value = original.get(key, value)
            child_path = (*_current_path, key)
            if key.lower() in _DEFAULT_SENSITIVE_KEYS:
                if _path_has_rule(rule_targeted_paths, child_path) or value != orig_value:
                    output[key] = value
                else:
                    output[key] = _REDACTED
                    if receipt_hook is not None:
                        joined = ".".join(cast(tuple[str, ...], child_path))  # pragma: no mutate — cast is identity
                        receipt_hook(
                            joined,
                            "redact",
                            orig_value,
                        )
            elif isinstance(value, str) and (redacted := _redact_if_secret(value)) is not None:
                # Span-scoped: only the secret-looking runs are replaced, so
                # the message around them survives.
                output[key] = redacted
                if receipt_hook is not None:
                    joined = ".".join(cast(tuple[str, ...], child_path))  # pragma: no mutate — cast is identity
                    receipt_hook(
                        joined,
                        "redact",
                        value,
                    )
            else:
                output[key] = _apply_default_sensitive_key_redaction(
                    value,
                    orig_value,
                    rule_targeted_paths=rule_targeted_paths,
                    depth=depth + 1,
                    max_depth=max_depth,
                    receipt_hook=receipt_hook,
                    _current_path=child_path,
                )
        return output
    if (
        isinstance(node, list) and isinstance(original, list)
    ):  # pragma: no mutate — dual isinstance guard; both False branches are unreachable given upstream recursion contract
        result: list[Any] = []
        for item, orig in zip(
            node, original, strict=False
        ):  # pragma: no mutate — strict=False because original may have extra trailing items after truncation upstream
            if isinstance(item, str) and (redacted := _redact_if_secret(item)) is not None:
                result.append(redacted)
                if receipt_hook is not None:
                    receipt_hook("(list_item)", "redact", item)
            else:
                result.append(
                    _apply_default_sensitive_key_redaction(
                        item,
                        orig,
                        rule_targeted_paths=rule_targeted_paths,
                        depth=depth + 1,
                        max_depth=max_depth,
                        receipt_hook=receipt_hook,
                        _current_path=(*_current_path, "*"),
                    )
                )
        return result
    return node


def _collect_rule_paths(rules: tuple[PIIRule, ...]) -> frozenset[tuple[str, ...]]:
    """Collect the full paths that custom rules target."""
    return frozenset(rule.path for rule in rules if rule.path)


def sanitize_payload(
    payload: dict[str, Any], enabled: bool, max_depth: int = 8
) -> dict[
    str, Any
]:  # pragma: no mutate — default max_depth=8 is overridden by live runtime config at every call; default is cosmetic
    if not enabled:
        return dict(payload)
    # Snapshot hooks once to prevent TOCTOU races if they are replaced concurrently.
    receipt_hook = _receipt_hook
    classification_hook = _classification_hook
    policy_fn = _policy_hook
    # Fix 3a: Thread-safe snapshot of rules list to prevent RuntimeError from
    # concurrent replace_pii_rules() calls mutating the list during iteration.
    with _lock:
        rules_snapshot = list(_rules)
    # _apply_rule builds entirely new dict/list nodes at every level it traverses,
    # so a shallow top-level copy is sufficient — no deepcopy needed.
    cleaned: Any = dict(payload)
    if rules_snapshot:
        rules = tuple(rules_snapshot)
        for rule in rules:
            cleaned = _apply_rule(cleaned, rule, receipt_hook=receipt_hook)
        rule_targeted_paths = _collect_rule_paths(rules)
    else:
        rule_targeted_paths = frozenset()  # pragma: no mutate — None also accepted by callee
    cleaned = _apply_default_sensitive_key_redaction(
        cleaned, payload, rule_targeted_paths=rule_targeted_paths, max_depth=max_depth, receipt_hook=receipt_hook
    )
    if classification_hook is not None and isinstance(cleaned, dict):
        # typing.cast returns its second argument unchanged at runtime, so every
        # mutation of the type argument is provably equivalent.
        items = list(cast(Any, cleaned).items())  # pragma: no mutate — cast returns its argument unchanged at runtime
        for key, value in items:
            label = classification_hook(key, value)
            if label is not None:
                # The fallback is only ever compared against "drop" and the
                # mask actions, so any other string is provably equivalent.
                default_action = "pass"  # pragma: no mutate — only compared against 'drop' and the mask actions; any other string is equivalent
                action = policy_fn(label) if policy_fn is not None else default_action
                if action == "drop":
                    del cleaned[key]
                else:
                    cleaned[f"__{key}__class"] = label
                    if action in ("redact", "hash", "truncate") and value != _REDACTED:
                        mask_mode = cast(MaskMode, action)  # pragma: no mutate — runtime no-op
                        cleaned[key] = _mask(value, mask_mode, 8)
    if isinstance(cleaned, dict):
        return cleaned
    return {}


def reset_pii_rules_for_tests() -> None:
    global _classification_hook, _receipt_hook, _policy_hook
    replace_pii_rules([])
    with _lock:
        _custom_secret_patterns.clear()
    _classification_hook = None
    _receipt_hook = None
    _policy_hook = None
